internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,156 @@
 type InMemoryWriteBuffer struct {
-	// Holds the data to be written to GCS. At any time, Buffer holds the data for
-	// 2 GCS write calls.
-	buffer *bytes.Buffer
+	// Holds the incoming data from kernel write calls, currently being buffered
+	// for upload in the future.
+	current inMemoryBuffer
+
+	// Holds the data that is currently being uploaded asynchronously to GCS. Data
+	// in this buffer will persist until the entire data is successfully uploaded.
+	flushed inMemoryBuffer
+
+	// Holds the last offset to which data is written.
+	fileSize int64
+
+	// Holds the buffer's size (in bytes) to be allocated to currentBuffer and flushedBuffer.
+	bufferSize int64
 }
 
-// CreateInMemoryWriteBuffer creates a buffer with InMemoryWriteBuffer.buffer
-// set to nil. Memory is allocated to the buffer when the first write call comes.
-// This avoids unnecessarily bloating GCSFuse memory consumption.
-//
-// To allocate memory to the buffer, use WriteBuffer.InitializeBuffer.
-func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
-	b := &InMemoryWriteBuffer{}
+// CreateInMemoryWriteBuffer returns a new InMemoryWriteBuffer.
+// Memory is lazily allocated to current and flushed buffer on their first usage.
+func CreateInMemoryWriteBuffer(bufferSizeMB uint) (*InMemoryWriteBuffer, error) {
+	if bufferSizeMB == 0 {
+		return nil, fmt.Errorf(ZeroSizeBufferError)
+	}
+	b := &InMemoryWriteBuffer{
+		bufferSize: int64(bufferSizeMB * MiB),
+	}
 	// TODO: set mtime attribute.
-	return b
+	logger.Debugf("TODO: set mtime attribute.")
+	return b, nil
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+// Allocates memory to currentBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureCurrentBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
 	}
+	if b.current.buffer == nil {
+		b.current.buffer = make([]byte, 0, b.bufferSize)
+		b.current.offset.end = b.bufferSize
+	}
+	return nil
 }
 
-func (b *InMemoryWriteBuffer) WriteAt(data []byte, offset int64) error {
+// Allocates memory to flushedBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureFlushedBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
+	}
+	if b.flushed.buffer == nil {
+		b.flushed.buffer = make([]byte, 0, b.bufferSize)
+	}
 	return nil
 }
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, offset int64) error {
+	dataSize := int64(len(content))
+	receivedContentEndOffset := offset + dataSize
+	if dataSize == 0 {
+		return nil
+	}
+
+	// Ensure b.currentBuffer != nil
+	if err := b.ensureCurrentBuffer(); err != nil {
+		return err
+	}
+
+	// If receivedContentStartOffset is not in the range of offsets stored in current
+	// buffer, trigger temp file flow for writes.
+	if offset < b.current.offset.start ||
+		offset > b.current.offset.end {
```

Comment:
I've explicitly kept the last offset included in the condition because kernel writes are expected to come at multiples of MiB.

Sample write logs to validate the same:
I wrote hello, hi and ashmeen to a file and we can see kernel sent the old data again just to keep the write as a multiple of 4096.
```
time="09/11/2023 04:39:40.174953" severity=TRACE msg="fuse_debug: Op 0x000001fc        connection.go:416] <- WriteFile (inode 5, PID 0, handle 32, offset 0, 5 bytes)"
...
time="09/11/2023 04:39:42.002566" severity=TRACE msg="fuse_debug: Op 0x0000020c        connection.go:416] <- WriteFile (inode 5, PID 0, handle 33, offset 0, 7 bytes)"
...
time="09/11/2023 04:39:43.454277" severity=TRACE msg="fuse_debug: Op 0x0000021a        connection.go:416] <- WriteFile (inode 5, PID 0, handle 34, offset 0, 14 bytes)"
```

Reply:
I initially implemented this entire PR with the assumption that kernel sends only 1MiB content in a write that is why this was written in this way. Agreed that this logic would not have worked if we ever had to reuse this logic elsewhere.

I've updated the PR with the following:
-  to consider random writes if current offset goes beyond the range of offsets stored in current buffer and +1 buffer. 
- added the logic to keep on writing until we exhaust the content received (more than 1MiB possible).
- removed not enough space in buffer errors

Please take another look. Thanks!

Reply:
Sure, I'll open any new questions on the updated design then. Marking this thread closed.